jina/executors/decorators.py

```diff
@@ -365,3 +372,95 @@ def arg_wrapper(*args, **kwargs):
         return _batching(func)
     else:
         return _batching
+
+def batching_ranker_input(func: Callable[[Any], np.ndarray] = None,
+                         batch_size: Union[int, Callable] = None,
+                         num_batch: Optional[int] = None,
+                         slice_on: Union[int, List[int]] = 2,
+                         num_data: int = None) -> Any:
+    """Split the input of a function into small batches and call :func:`func` on each batch
+    , collect the merged result and return. This is useful when the input is too big to fit into memory.
+    This can batch multiple inputs. 
+
+    :param func: function to decorate
+    :param batch_size: size of each batch
+    :param num_batch: number of batches to take, the rest will be ignored
+    :param slice_on: the location of the data. When using inside a class,
+            ``slice_on`` should take ``self`` into consideration.
+    :param num_data: the number of data inside the arguments
+    :return: the merged result as if run :func:`func` once on the input.
+    ..warning:
+        data arguments will be taken starting from ``slice_on` to ``slice_on + num_data``
+    Example:
+        .. highlight:: python
+        .. code-block:: python
+            
+            class MemoryHungryRanker:
+                
+                @batching_ranker_input(batch_size = 64)
+                def score(
+                    self, query_meta: Dict, old_match_scores: Dict, match_meta: Dict
+                ) -> 'np.ndarray': 
+                ...       
+
+                @batching_ranker_input(batch_size = 64, num_data = 1)
+                def score(
+                    self, query_meta: Dict, old_match_scores: Dict, match_meta: Dict
+                ) -> 'np.ndarray': 
+                ...
+                    
+    """
+    def _batching(func):
+        @wraps(func)
+        def arg_wrapper(*args, **kwargs):
+            _slice_on = slice_on
+            _num_data = num_data
+            if _num_data is not None:
+                if isinstance(_slice_on, List):
+                    raise ValueError(f'When using num_data in @batching_ranker_input, an integer value '
```

Comment:
then please fix the `slice_on` type hint, since it suggests a `List` is accepted

Reply:
I would just adapt the `type hint` and remove this check

Reply:
Avoid this noisy part, consider num_data is never None (type hint suggests so)

Reply:
I was thinking, `slice_on` and `num_data` when used together, allow you to batch consecutive parameters. There might be a case where you want to batch non - consecutive parameters, lets say [1,3], in those cases we can have `slice_on` as list of indices. So whenever  `slice_on` is passed as a list, `num_data` (type hint needs to be updated to optional) cannot be used. wdyt?

Reply:
Since there is not yet a case where we require this, I'd prefer to keep it simple